lib/src/rules/type_annotate_public_apis.dart

```diff
@@ -75,14 +75,24 @@ class _Visitor extends SimpleAstVisitor<void> {
   _Visitor(this.rule) : v = new _VisitorHelper(rule);
   @override
   void visitFieldDeclaration(FieldDeclaration node) {
-    if (node.fields.type == null) {
+    if (node.fields.type == null && !_isParentClassOrMixinPrivate(node)) {
+      // This is an untyped field in a public class or mixin.
+
+      // kevmoo@ points out that a private class with a public field may be
+      // extended by a public class which inherits the private class's
+      // implementation of that field, in which case it becomes public API.
+      // TODO(srawlins): Add logic checking "is this field somehow inherited in
+      // a public class"?
```

Comment:
We don't actually have the ability to check for public subclasses. Doing so requires a closed world assumption, and analyzer doesn't support that assumption.

It isn't clear to me that we want to ignore members in private classes, but if we do, then we won't be able to fix the case that Kevin pointed out.

Reply:
Really? How can someone define a public class extending a private one, but that couldn't be visited by the same linter rule visitor?

Reply:
You're partially right. I was definitely forgetting that we're talking about private classes that won't be visible outside the library. That said, the visitor won't see code from other parts.

Reply:
Ahh, good point.

Hmm, yeah this is an interesting lint rule. Nowhere is "public API" defined ☹️.

I would be super surprised to receive a lint report that I should document a public member of a private class as "public API", but we actually don't have a bug report on that 😛.

I could punt on the private stuff, and just implement the fix for #1108 here regarding functions-within-functions. WDYT?

Reply:
> ... just implement the fix for #1108 here regarding functions-within-functions.

Sounds good to me.

> Nowhere is "public API" defined ...

Actually, it's defined in two ways in two places, which is sort of equivalent. :-)

The spec says:

> A declaration is _private_ iff its name is _private_, otherwise it is _public_.

While `pub` defines what we mean by a public library (https://www.dartlang.org/tools/pub/package-layout#public-directories).

I don't know which definition we want to use here.

Reply:
Ok undone for now.